File: 2-d-materiais-pedagogicos/material-2011-Java-prototipo-experimental-MPD-orm-queries/orm-queries/orm-queries/src/main/java/northwind/model/Product.java
package northwind.model;

import orm.annotations.JdbcCol;
import orm.annotations.JdbcMapper;
import orm.types.ValueHolder;

import java.sql.SQLException;
import orm.Entity;

/**
 * @author  mcarvalho
 */
public class Product implements Entity<Integer>{
	
	@JdbcMapper(table="Products")
	public Product(
			@JdbcCol(value = "ProductId", isPk=true, isIdentity = true) int productID, 
			@JdbcCol("ProductName") String productName, 
			@JdbcCol("UnitPrice") double unitPrice,
			@JdbcCol("UnitsInStock") int unitsInStock,
			@JdbcCol(value = "SupplierId", referencedKeyClass=Integer.class) ValueHolder<Supplier> supplier,
			@JdbcCol(value = "ProductId") Iterable<OrderDetails> orders,
			@JdbcCol(value = "CategoryID", referencedKeyClass=Integer.class) ValueHolder<Category> category) {
		super();
		this.productID = productID;
		this.productName = productName;
		this.unitPrice = unitPrice;
		this.unitsInStock = unitsInStock;
		this.supplier = supplier;
		this.orders = orders;
		this.category = category;
	}
	public Product(String productName, double unitPrice, int unitsInStock) {
		this.productName = productName;
		this.unitPrice = unitPrice;
		this.unitsInStock = unitsInStock;
	}
	int productID;
	String productName;
	double unitPrice;
	int unitsInStock;
	ValueHolder<Supplier> supplier;
	int quantityPerUnit;
	int unitsOnOrder;
	int reorderLevel;
	boolean discontinued;
	Iterable<OrderDetails> orders;
	ValueHolder<Category> category;
	
	
	public Category getCategory() {
		return category.get();
	}
	public void setCategory(final Category category) {
		this.category = new ValueHolder<Category>() {
			@Override
			public Category get() {
				return category;
			}
			
		};
	}
	public Iterable<OrderDetails> getOrders() {
		return orders;
	}
	public void setOrders(Iterable<OrderDetails> orders) {
		this.orders = orders;
	}
	public Integer getId() {
		return productID;
	}
	public void setId(Integer productID) {
		this.productID = productID;
	}

	public String getProductName() {
		return productName;
	}

	public void setProductName(String productName) {
		this.productName = productName;
	}

	public double getUnitPrice() {
		return unitPrice;
	}

	public void setUnitPrice(double unitPrice) {
		this.unitPrice = unitPrice;
	}

	public int getUnitsInStock() {
		return unitsInStock;
	}

	public void setUnitsInStock(int unitsInStock) {
		this.unitsInStock = unitsInStock;
	}

	public Supplier getSupplier() throws SQLException{
		return supplier.get();
	}

	public void setSupplier(final Supplier s){
		supplier = new ValueHolder<Supplier>(){
			@Override
			public Supplier get(){
				return s;
			}
		};
	}
	@Override
	public String toString() {
		return "Product [productID=" + productID + ", productName="
				+ productName + ", unitPrice=" + unitPrice + ", unitsInStock="
				+ unitsInStock + ", supplier=" + supplier.get().getId() + ", orders="
				+ orders + "]";
	}
}

File: 2-d-materiais-pedagogicos/material-2013-Dotnet-prototipo-experimental-AVE-sql-mapper/SqlMapper/src/SqlMapperNorthwind/Model/Product.cs
using System;

namespace SqlMapperNorthwind.Model
{
    public class Product : IEquatable<Product>
    {
        public int ProductID { set; get; }
        public string ProductName { set; get; }
        public int SupplierID { set; get; }
        public int CategoryID { set; get; }
        public string QuantityPerUnit { set; get; }
        public decimal UnitPrice { set; get; }
        public short UnitsInStock { set; get; }
        public short UnitsOnOrder { set; get; }
        public short ReorderLevel { set; get; }
        public bool Discontinued { set; get; }

        public Product()
        {
            
        }

        public Product(
            int categoryId, 
            bool discontinued, 
            int productId, 
            string productName,
            string quantityPerUnit, 
            short reorderLevel, 
            int supplierId, 
            decimal unitPrice,
            short unitsInStock, 
            short unitsOnOrder)
        {
            CategoryID = categoryId;
            Discontinued = discontinued;
            ProductID = productId;
            ProductName = productName;
            QuantityPerUnit = quantityPerUnit;
            ReorderLevel = reorderLevel;
            SupplierID = supplierId;
            UnitPrice = unitPrice;
            UnitsInStock = unitsInStock;
            UnitsOnOrder = unitsOnOrder;
        }

        public bool Equals(Product other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return ProductID == other.ProductID && string.Equals(ProductName, other.ProductName) && SupplierID == other.SupplierID && CategoryID == other.CategoryID && string.Equals(QuantityPerUnit, other.QuantityPerUnit) && UnitPrice == other.UnitPrice && UnitsInStock == other.UnitsInStock && UnitsOnOrder == other.UnitsOnOrder && ReorderLevel == other.ReorderLevel && Discontinued.Equals(other.Discontinued);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Product)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hashCode = ProductID;
                hashCode = (hashCode * 397) ^ (ProductName != null ? ProductName.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ SupplierID;
                hashCode = (hashCode * 397) ^ CategoryID;
                hashCode = (hashCode * 397) ^ (QuantityPerUnit != null ? QuantityPerUnit.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ UnitPrice.GetHashCode();
                hashCode = (hashCode * 397) ^ UnitsInStock.GetHashCode();
                hashCode = (hashCode * 397) ^ UnitsOnOrder.GetHashCode();
                hashCode = (hashCode * 397) ^ ReorderLevel.GetHashCode();
                hashCode = (hashCode * 397) ^ Discontinued.GetHashCode();
                return hashCode;
            }
        }


    public override string ToString()
        {
            return "Product [id=" + ProductID + ", name=" + ProductName + ", price=" + UnitPrice
                    + ", stock=" + UnitsInStock + "]";
        }
    }
}
